pygac/reader.py

```diff
@@ -116,38 +127,181 @@ def __init__(self, interpolate_coords=True, adjust_clock_drift=True,
-            This method should be called inside the Reader.read implementation
+            This is a classmethod to avoid throwaway instances while
+            checking if the reader corresponds to the input file.
         """
-        basename = os.path.basename(filename)
-        root, ext = os.path.splitext(basename)
-        if ext == ".gz":
-            self.filename = root
-            file_object = gzip.open(filename, mode='rb')
-        else:
-            self.filename = basename
-            file_object = open(filename, mode='rb')
-        LOG.info('Reading %s', self.filename)
+        raise NotImplementedError
+
+    @classmethod
+    def _validate_header(cls, header):
+        """Check if the header belongs to this reader
+
+        Note:
+            according to https://www1.ncdc.noaa.gov/pub/data/satellite/
+            publications/podguides/TIROS-N%20thru%20N-14/pdf/NCDCPOD2.pdf
+            and https://www1.ncdc.noaa.gov/pub/data/satellite/
+            publications/podguides/N-15%20thru%20N-19/pdf/
+            2.5%20Section%208.0%20NOAA%20Level%201B%20Database.pdf
+            the data set name splits into
+            PROCESSING-CENTER.DATA-TYPE.SPACECRAFT-UNIQUE-ID.
+            YEAR-DAY.START-TIME.STOP-TIME.PROCESSING-BLOCK-ID.SOURCE
+            This should be sufficient information to determine the
+            reader.
+            Global Area Coverage (GAC):
+                DATA-TYPE = GHRR
+            Local Area Coverage (LAC):
+                DATA-TYPE = LHRR
+            Polar Orbiter Data (POD):
+                SPACECRAFT-UNIQUE-ID in [TN, NA, NB, NC, ND, NE, NF,
+                                         NG, NH, NI, NJ]
+            NOAA-K, -L, -M system, but also newer satellites (KLM):
+                SPACECRAFT-UNIQUE-ID in [NK, NL, NM, NN, NP, M2, M1]
+        """
+        # This method does not need to be implemented in all subclasses.
+        # It is intended for cooperative multiple inheritance, i.e.
+        # each child class which implements this method, should call the
+        # super method to enter into the method resolution order.
+        # See https://docs.python.org/3/library/functions.html#super
+        # second use case “diamond diagrams”.
+        # Check if the data set name matches the pattern
+        LOG.debug("validate header")
+        try:
+            data_set_name = header['data_set_name'].decode()
+        except UnicodeDecodeError:
+            raise ReaderError('Not able to decode the data set name!')
+        if not cls.data_set_pattern.match(data_set_name):
+            raise ReaderError(
+                'Data set name "%s" does not match!' % data_set_name)
+
+    def _read_scanlines(self, buffer, count):
+        """Read the scanlines from the given buffer
+
+        Args:
+            buffer (bytes, bytearray): buffer to read from
+            count (int): number of expected scanlines
+        """
+        line_count = len(buffer) // self.scanline_type.itemsize
```

Comment:
Maybe add a comment here, that it is possible to have an incomplete scanline in the end of a file which will be removed by this method?